common/cache/budget.go

```diff
@@ -0,0 +1,1260 @@
+//   - softCapThreshold: optional percentage (0.0-1.0) defining how capacity is split between two tiers:
+//   - Free space tier: (threshold * capacity) - shared by all caches
+//   - Fair share tier: ((1 - threshold) * capacity) / activeCaches - allocated per cache
+//     The fair share capacity is equal to the fair share capacity ((1 - threshold) * capacity) divided by the number
+//     of active caches (caches with usage > 0).
+func NewBudgetManager(
+	name string,
+	maxBytes dynamicproperties.IntPropertyFn,
+	maxCount dynamicproperties.IntPropertyFn,
+	admission AdmissionMode,
+	reclaimBackoff time.Duration,
+	scope metrics.Scope,
+	logger log.Logger,
+	softCapThreshold dynamicproperties.FloatPropertyFn, // nil defaults to 1.0 (disabled)
+) Manager {
+	if scope == nil {
+		scope = metrics.NoopScope
+	}
+	if maxBytes == nil {
+		maxBytes = dynamicproperties.GetIntPropertyFn(-1) // unlimited if maxBytes is not provided
+	}
+	if maxCount == nil {
+		maxCount = dynamicproperties.GetIntPropertyFn(-1) // unlimited if maxCount is not provided
+	}
+	if softCapThreshold == nil {
+		softCapThreshold = dynamicproperties.GetFloatPropertyFn(1.0) // no threshold if softCapThreshold is not provided
+	}
+	if admission == 0 {
+		admission = AdmissionOptimistic
+	}
+
+	mgr := &manager{
+		name:                        name,
+		maxBytes:                    maxBytes,
+		maxCount:                    maxCount,
+		reclaimBackoff:              reclaimBackoff,
+		admission:                   admission,
+		scope:                       scope.Tagged(metrics.BudgetManagerNameTag(name)),
+		logger:                      logger,
+		enforcementSoftCapThreshold: softCapThreshold,
+	}
+
+	// Emit initial metrics
+	mgr.updateMetrics()
+
+	return mgr
+}
+
+// updateMetrics emits current state metrics
+func (m *manager) updateMetrics() {
+	// Emit capacity metrics
+	capacityBytes := m.CapacityBytes()
+	if capacityBytes != math.MaxUint64 {
+		m.scope.UpdateGauge(metrics.BudgetManagerCapacityBytes, float64(capacityBytes))
+	}
+
+	capacityCount := m.CapacityCount()
+	if capacityCount != math.MaxInt64 {
+		m.scope.UpdateGauge(metrics.BudgetManagerCapacityCount, float64(capacityCount))
+	}
+
+	// Emit usage metrics
+	m.scope.UpdateGauge(metrics.BudgetManagerUsedBytes, float64(m.UsedBytes()))
+	m.scope.UpdateGauge(metrics.BudgetManagerUsedCount, float64(m.UsedCount()))
+
+	// Emit soft threshold
+	m.scope.UpdateGauge(metrics.BudgetManagerSoftThreshold, m.enforcementSoftCapThreshold())
+
+	// Emit active cache count
+	m.scope.UpdateGauge(metrics.BudgetManagerActiveCacheCount, float64(atomic.LoadInt64(&m.activeCacheCount)))
+}
+
+// emitHardCapExceeded logs when hard capacity limit is exceeded and increments counter
+func (m *manager) emitHardCapExceeded(cacheID, budgetType string, requested, available uint64) {
+	if m.scope != nil {
+		m.scope.IncCounter(metrics.BudgetManagerHardCapExceeded)
+	}
+
+	if m.logger != nil {
+		m.logger.Debug("Hard capacity limit exceeded",
```

Comment:
I guess we should make this consistent with the soft-cap, either they both should be debug or warn?

Reply:
discussed offline, may move to a throttled logger in a future PR